services/server/server.go

```diff
@@ -110,87 +111,39 @@ func (s *Server) UpdateSettingsFromEnv(env []string) error {
 	defer s.envRW.Unlock()
 
+	envSettings, errs, warns := validators.ValidateEnvVars(env)
+	for _, w := range warns {
+		s.l.Warnln(w)
+	}
+
+	// This should be impossible in the normal workflow.
+	// An invalid environment variable must be caught with pmm-managed-init
+	// and the docker run must be terminated.
+	if len(errs) > 0 {
+		for _, e := range errs {
+			s.l.Errorln(e)
+		}
+		return errors.New("validation error of environment variables")
+	}
+
+	s.envDisableUpdates = envSettings.DisableUpdates
+	s.envDisableTelemetry = envSettings.DisableTelemetry
+	s.envMetricsResolutionHR = envSettings.MetricsResolutions.HR
+	s.envMetricsResolutionMR = envSettings.MetricsResolutions.MR
+	s.envMetricsResolutionLR = envSettings.MetricsResolutions.LR
+	s.envDataRetention = envSettings.DataRetention
+
 	return s.db.InTransaction(func(tx *reform.TX) error {
 		settings, err := models.GetSettings(tx.Querier)
 		if err != nil {
 			return err
 		}
 
-		for _, e := range env {
-			p := strings.SplitN(e, "=", 2)
-			if len(p) != 2 {
-				s.l.Warnf("Failed to parse environment variable %q.", e)
-				continue
-			}
-
-			k, v := strings.ToUpper(p[0]), strings.ToLower(p[1])
-			var err error
-			var value interface{}
-			switch k {
-			case "DISABLE_UPDATES":
-				var b bool
-				b, err = strconv.ParseBool(v)
-				if err == nil {
-					value = b
-					s.envDisableUpdates = b
-				}
-
-			case "DISABLE_TELEMETRY":
-				var b bool
-				b, err = strconv.ParseBool(v)
-				if err == nil {
-					value = b
-					s.envDisableTelemetry = b
-					settings.Telemetry.Disabled = b
-				}
-
-			case "METRICS_RESOLUTION", "METRICS_RESOLUTION_HR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionHR = d
-					settings.MetricsResolutions.HR = d
-				}
-
-			case "METRICS_RESOLUTION_MR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionMR = d
-					settings.MetricsResolutions.MR = d
-				}
-
-			case "METRICS_RESOLUTION_LR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionLR = d
-					settings.MetricsResolutions.LR = d
-				}
-
-			case "DATA_RETENTION":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envDataRetention = d
-					settings.DataRetention = d
-				}
-
-			default:
-				s.l.Tracef("Skipping %q.", e)
-				continue
-			}
-
-			if err == nil {
-				s.l.Infof("Environment variable %q parsed: %v.", e, value)
-			} else {
-				s.l.Warnf("Failed to parse environment variable %q: %s.", e, err)
-			}
-		}
+		settings.Telemetry.Disabled = s.envDisableTelemetry
+		settings.MetricsResolutions.HR = s.envMetricsResolutionHR
+		settings.MetricsResolutions.MR = s.envMetricsResolutionMR
+		settings.MetricsResolutions.LR = s.envMetricsResolutionLR
+		settings.DataRetention = s.envDataRetention
```

Comment:
So we always overwrite settings with values from environment variables, even if they are not provided, have zero values, and then filled with default values on save? So we always lose settings on pmm-managed restart?

Reply:
good point, missed that